x/sanction/keeper/keeper.go

```diff
@@ -0,0 +1,408 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	"github.com/cosmos/cosmos-sdk/x/sanction"
+	"github.com/cosmos/cosmos-sdk/x/sanction/errors"
+)
+
+type Keeper struct {
+	cdc      codec.BinaryCodec
+	storeKey storetypes.StoreKey
+
+	govKeeper sanction.GovKeeper
+
+	authority string
+
+	unsanctionableAddrs map[string]bool
+
+	msgSanctionTypeURL          string
+	msgUnsanctionTypeURL        string
+	msgExecLegacyContentTypeURL string
+}
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey storetypes.StoreKey,
+	bankKeeper sanction.BankKeeper,
+	govKeeper sanction.GovKeeper,
+	authority string,
+	unsanctionableAddrs []sdk.AccAddress,
+) Keeper {
+	rv := Keeper{
+		cdc:                         cdc,
+		storeKey:                    storeKey,
+		govKeeper:                   govKeeper,
+		authority:                   authority,
+		unsanctionableAddrs:         make(map[string]bool),
+		msgSanctionTypeURL:          sdk.MsgTypeURL(&sanction.MsgSanction{}),
+		msgUnsanctionTypeURL:        sdk.MsgTypeURL(&sanction.MsgUnsanction{}),
+		msgExecLegacyContentTypeURL: sdk.MsgTypeURL(&govv1.MsgExecLegacyContent{}),
+	}
+	for _, addr := range unsanctionableAddrs {
+		// using string(addr) here instead of addr.String() to cut down on the need to bech32 encode things.
+		rv.unsanctionableAddrs[string(addr)] = true
```

Comment:
I find this interesting.  What you say here makes sense to cut down on the bech32 operation, but would there ever be a case we would want to query this map and get the bech32 addresses keys?  I'm not sure if there is an easy way to convert a string(sdk.AccAddress) back to an sdk.AccAddress.

Reply:
To get the bech32 from one of these keys, you'd cast the key back to an `sdk.AccAddress` and `.String()` it, e.g. `sdk.AccAddress(key).String()`. To go from a bech32 string back to an `sdk.AccAddress`, the function`sdk.AccAddressFromBech32` is used. The `sdk.AccAddress` is defined as `type AccAddress []byte` which is why it can be cast to a string like I'm doing here, and cast back if needed. It's not going to be a printable string, but that doesn't matter here.

I'm under the impression that neither `sdk.AccAddress#String` nor `sdk.AccAddressFromBech32` are trivial when it comes to processing, and that their use should be minimized. I'm not really sure how much processing they require though. My impression comes from some comments (I think in the bank module) about minimizing their use. I felt it prudent to not `.String()` it since A) `IsSanctioned` takes in an `AccAddress`, B) The `IsSanctioned` checking doesn't ever need the bech32, and C) `IsSanctioned` is called during `SendCoins`, which is used enough that skipping a -> bech32 conversion might make a difference.